File: strings/sliding_window/permutations_in_a_string.rb
def check_inclusion(p, s)
    return false if s.size < p.size
    m, n = s.size, p.size
    curr, desired = [0] * 26, [0] * 26
    count = 0
    
    0.upto(p.size - 1) do |i|
        curr[s[i].ord - 'a'.ord] += 1
        desired[p[i].ord - 'a'.ord] += 1
    end
    0.upto(25) do |i|
        count += 1 if curr[i] == desired[i]
    end
    return true if count == 26

    1.upto(m - n) do |i|
        l = s[i - 1].ord - 'a'.ord
        r = s[i - 1 + n].ord - 'a'.ord
        count -= 1 if curr[l] == desired[l]
        curr[l] -= 1
        count += 1 if curr[l] == desired[l]

        count -= 1 if curr[r] == desired[r]
        curr[r] += 1
        count += 1 if curr[r] == desired[r]
        
        return true if count == 26
    end

    false
end


# 567. Permutation in String
# https://leetcode.com/problems/permutation-in-string/description/

File: strings/sliding_window/permutations_in_a_string.py
import collections

class Solution1:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if not s1 or len(s1) > len(s2): return False

        s1_hash = collections.Counter(s1)
        m, n, desired, formed = len(s1), len(s2), len(s1_hash), 0
        s_hash = collections.defaultdict(int)

        for j in range(m):
            c = s2[j]
            s_hash[c] += 1
            if c in s1_hash and s1_hash[c] == s_hash[c]: formed += 1
        if formed == desired: return True

        r = m
        while r < n:
            l = r - m + 1
            out_c, in_c = s2[l - 1], s2[r]

            if out_c in s1_hash and s1_hash[out_c] == s_hash[out_c]: formed -= 1
            s_hash[out_c] -= 1

            s_hash[in_c] += 1
            if in_c in s1_hash and s1_hash[in_c] == s_hash[in_c]: formed += 1

            if formed == desired: return True
            r += 1

        return False


class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        m, n = len(s1), len(s2)
        d1, d2 = collections.Counter(s1), collections.Counter(s2[:m])
        if d1 == d2: return True

        for i in range(m, n):
            out_c, in_c = s2[i - m], s2[i]
            d2[out_c] -= 1; d2[in_c] += 1
            if d2[out_c] == 0: d2.pop(out_c)
            if d1 == d2: return True

        return False




# 567. Permutation in String
# https://leetcode.com/problems/permutation-in-string/description/
